src/writers/sink.creator.cpp

```diff
@@ -0,0 +1,409 @@
+#include "sink.creator.hh"
+#include "file.sink.hh"
+
+#include <latch>
+#include <queue>
+
+namespace zarr = acquire::sink::zarr;
+namespace common = zarr::common;
+
+zarr::SinkCreator::SinkCreator(std::shared_ptr<common::ThreadPool> thread_pool_)
+  : thread_pool_{ thread_pool_ }
+{
+}
+
+bool
+zarr::SinkCreator::create_chunk_sinks(
+  const std::string& base_uri,
+  const std::vector<Dimension>& dimensions,
+  std::vector<std::shared_ptr<Sink>>& chunk_sinks)
+{
+    return make_part_sinks_(
+      base_uri, dimensions, common::chunks_along_dimension, chunk_sinks);
+}
+
+bool
+zarr::SinkCreator::create_shard_sinks(
+  const std::string& base_uri,
+  const std::vector<Dimension>& dimensions,
+  std::vector<std::shared_ptr<Sink>>& shard_sinks)
+{
+    return make_part_sinks_(
+      base_uri, dimensions, common::shards_along_dimension, shard_sinks);
+}
+
+bool
+zarr::SinkCreator::create_v2_metadata_sinks(
+  const std::string& base_uri,
+  size_t n_arrays,
+  std::vector<std::shared_ptr<Sink>>& metadata_sinks)
+{
+    std::queue<std::string> dir_paths, file_paths;
+
+    file_paths.emplace(".metadata"); // base metadata
```

Comment:
I'm going to keep this as it is for now, because in a future PR I want to change these to a key-value store. It turns out the order in which these paths are stored is important to the way they're used elsewhere, such that if one were to add something here, it would silently break the correct placement of the metadata.